packages/dds/tree/src/shared-tree/transactionTypes.ts

```diff
@@ -0,0 +1,160 @@
+	/** Property indicating that the transaction failed */
+	readonly success: false;
+}
+
+/**
+ * The extended successful return value of the runTransaction API, i.e., the transaction succeeded.
+ * @alpha
+ */
+export interface RunTransactionSucceededExt<TSuccessValue> extends RunTransactionSucceeded {
+	/** The user defined return value on successfully completing the transaction */
+	readonly returnValue: TSuccessValue;
+}
+/**
+ * The extended failed return value of the runTransaction API, i.e., the transaction failed.
+ * @alpha
+ */
+export interface RunTransactionFailedExt<TFailureValue> extends RunTransactionFailed {
+	/** The user defined return value on failing the transaction */
+	readonly returnValue: TFailureValue;
+}
+
+/**
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface RunTransactionParams {
+	/**
+	 * The function to run as the body of the transaction.
+	 * @returns The result of the transaction. The user provided result (TResult) can either be returned directly or
+	 * as part of the `TransactionOutcome` object which can include other properties.
+	 * It could return nothing (TResult == void) to indicate a successful transaction.
+	 *
+	 * At any point during the transaction, the function may return the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) to abort the transaction and discard any changes it made so far.
+	 */
+	readonly transaction: () => ContinueTransaction | AbortTransaction;
+	/**
+	 * An optional list of {@link TransactionConstraint | constraints} that are checked just before the transaction begins.
+	 * If any of the constraints are not met when `runTransaction` is called, it will throw an error.
+	 * If any of the constraints are not met after the transaction has been ordered by the service, it will be rolled back on
+	 * this client and ignored by all other clients.
+	 */
+	readonly preconditions?: readonly TransactionConstraint[];
+}
+
+/**
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface RunTransactionParamsExt<TSuccessValue, TFailureValue> {
+	/**
+	 * The function to run as the body of the transaction.
+	 * @returns The result of the transaction. The user provided result (TResult) can either be returned directly or
+	 * as part of the `TransactionOutcome` object which can include other properties.
+	 * It could return nothing (TResult == void) to indicate a successful transaction.
+	 *
+	 * At any point during the transaction, the function may return the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) to abort the transaction and discard any changes it made so far.
+	 */
+	readonly transaction: () =>
+		| ContinueTransactionExt<TSuccessValue>
+		| AbortTransactionExt<TFailureValue>;
```

Comment:
We'd be supporting `ContinueTransaction | AbortTransaction` and `ContinueTransactionExt<TSuccessValue> | AbortTransactionExt<TFailureValue>` but not `ContinueTransaction | AbortTransactionExt<TFailureValue>` or `ContinueTransactionExt<TSuccessValue> | AbortTransaction`.

I'm cool with that, but I wanted to call it out, to make sure this is a conscious choice.

Reply:
I am fine with that too. We can always add it if needed. We could still support it because the implementation wouldn't change. However, it will add more types and make it complex.